### ISSUE TITLE
Allow forward and reverse history searches in the console

### ISSUE THREAD

**Issue by Zaeph** [ranger/ranger#1292]
#### RUNTIME ENVIRONMENT

- Operating system and version: Arch Linux x86_64
- Terminal emulator and version: terminator 1.91
- Python version: 3.7.0 (default, Jul 15 2018, 10:44:58) [GCC 8.1.1 20180531]
- Ranger version/commit: ranger-master 1.9.1
- Locale: En/GB

#### Feature Suggestion
The `console` in `ranger` doesn’t allow for reverse and forward history searches which is implemented in most terminals via the following readline commands:
- `C-r` for reverse search,
- `C-s` for forward search,
- `Alt-r` for reverse pattern search,
- `Alt-s` for forward pattern search.

After browsing the issue tracker, I’ve come to understand that most things in `ranger` should be achieved via key-bindings, and that the use of the `console` should be kept low. Nevertheless, I often find myself wishing that I could use the ability to select files in `ranger` with the full-fledged capabilities of my terminal such as:
- **Easy movement** between arguments,
- **`Tab`-completion** for commands,
- A **searchable history**.

To some extent, some of those aspects have already been implemented. For instance, I’ve made an attempt in #1285 to restore movements by words in `ranger`’s `console`, and it’s already proven quite useful to me. Other aspects such as `Tab`-completion are only going to take a bit of tweaking (I’ve discussed this in #706), but the remaining ones are likely to be the real challenges, at least to my untrained eye.

All in all, the four commands would add a decent amount of flexibility to the `console` for menial tasks. Again, because people are likely to drop into a full-fledged shell for complex operations, the goal here would merely be to keep an easily-searchable catalogue of basic commands which would benefit from `ranger`’s breezy selection.

#### Discussion on the role that the `console` in `ranger` should have

I’d like to hear if I’m the only one interested in overhauling the `console` in `ranger`, or if people are keener on dropping into a full-fledged shell even for trivial tasks.
I'll readily admit that globbing files inside a full-fledged shell is suitable, but perhaps a bit expedient when it requires us to bulk-rename the files beforehand. To my mind, improving the `console` in `ranger` has the potential to cover this grey area.

**Comment by Zaeph** [comment]
That would indeed be a nice implementation too.

I was puzzled at first because I thought I’d had mentioned this amongst the 4 commands, but I now realised that I’ve mixed things up in my report. The behaviour you describe is what I referred to as *pattern search* which wasn’t very descriptive, and the key-bindings were wrong, too. I’ve edited the issue to reflect this.

In `terminator`, the key-bindings for the behaviour you describe are `Alt-p` and `Alt-n`, and judging by `rc.conf`, they seem to be available.

I’m not sure exactly how those 4 commands are implemented in our terminals, but since we already have a file for the history, I guess it’s only a matter of writing functions to handle the searches and providing mappings for the console with `cmap`.

**Comment by toonn** [comment]
Hmm, `alt-p` doesn't seem to resemble fish's features at all. It seems to give me the first match for the string I give. That's imo way less useful than the interactive versions but I guess we could have them for completeness.

**Comment by toonn** [comment]
Btw, I think the behavior is already present in ranger so it shouldn't be hard to add. Try `:<some prefix of something you have in your history>` then press `<Up>`.

**Comment by Zaeph** [comment]
You’re absolutely right, it only preserves the first string in the command. Sorry for jumping to conclusions to quickly.

So, that means that out of the 4 commands I mentioned, 3 and 4 are already implemented, and can be accessed either with `<UP>`, `<DOWN>`, or their `copycmap` alternatives `Ctrl-p` and `Ctrl-n`. That’s really good, and now I want my terminal to do the same because it’s pretty useful. :)

I’m not sure if it could entirely replace searching through the history, but I’d say it covers most of what we could possibly need (at least to my tired mind).

**Comment by toonn** [comment]
Not quite sure what "preserves the first string in the command" means.
This is what it does for me:
 1. `$ (empty prompt) # alt-p`
 2. `$ :`
 3. `$ :Dow<enter>`
 4. `$ ls Downloads/`

**Comment by Zaeph** [comment]
Sorry, I don’t think I was thinking clearly yesterday.

### Tests
Here’s the differences I have found between `Ctrl-p` and `Alt-p` in `terminator` with `zsh`+`prezto` :
1. `Ctrl-p` in `terminator` with `zsh`+`prezto`
```sh
$ exec zsh
$ (empty prompt)
$ ls ~/Documents
$ touch test_file
$ ls ~/Downloads
$ (empty prompt)   # Ctrl-p
$ ls ~/Downloads   # Ctrl-p
$ touch test_file  # Ctrl-p  <<
$ ls ~/Documents
```
2. `Alt-p` in `terminator` with `zsh`+`prezto`
```sh
$ exec zsh
$ (empty prompt)
$ ls ~/Documents
$ touch test_file
$ ls ~/Downloads
$ ls		   # Alt-p
$ ls ~/Downloads   # Alt-p
$ ls ~/Documents
```
```sh
$ exec zsh
$ (empty prompt)
$ ls ~/Documents
$ ls ~/Downloads
$ ls ~/Images
$ ls ~/Do	   # Alt-p
$ ls ~/Images	   # Alt-p  <<  [1]
$ ls ~/Downloads   # Alt-p
$ ls ~/Documents
```

Then, I installed `fish` to check out the feature you described:
- Note that `Alt-p` has been replaced with `<UP>`.
  `Alt-p` added as string to the prompt, but I couldn’t figure out what it was and from where it came. All I know is that it didn’t seem to work on the command history.

3. `<UP>` in `terminator` with `fish`
```sh
$ exec fish
$ (empty prompt)
$ ls ~/Documents
$ touch test_file
$ ls ~/Downloads
$ ls		   # <UP>
$ ls ~/Downloads   # <UP>
$ ls ~/Documents
```
```sh
$ exec fish
$ (empty prompt)
$ ls ~/Documents
$ ls ~/Downloads
$ ls ~/Images
$ ls ~/Do	   # <UP>
$ ls ~/Downloads   # <UP>  <<
$ ls ~/Documents
```

### Observations :
- **`Ctrl-p` and `Alt-p` have different behaviours in `zsh`,**
  - `Alt-p`’s behaviour is what I referred to as a search which ‘preserves the first string in the command’, but a better description would be that it *searches the history based on the **first atomic string** found in the prompt* (cf. [1]),
- **`Alt-p` in `zsh` and `<UP>` in `fish` have comparable aims, but `fish`’s implementation seems more complete and robust**.
  The fact that the string being searched (i.e. from beginning of the prompt to the point) is highlighted is also a nice feature, and so is the preview of potential arguments based on the command history.

I hope this clarifies your question. Looking into it piqued my interest, so I might have a more thorough look at `fish`.

At any rate, for the matter at hand, and since we have the choice to implement whatever feature we want without being bound to a given shell, I’d rather have `fish`’s implementation rather than `zsh`’s since, to me, the former seems to be a super-set of the later.

**Comment by toonn** [comment]
Yeah, I really like fish's history completion. There's plugins for zsh that mimic it but there's nothing that achieves feature parity. It's fish's killer feature imo. The other one is that I'm still using a completely default configuration for fish, yes the defaults are just that good.

I have zsh on two machines and the configs are different even though I went through the configuration helper tool with both of them not long apart. And I don't like either configuration.

I do have a couple functions for fish and I have changed the PATH a couple times of course. But I think I haven't even customized the PS1 for fish which definitely was a pet peeve of mine with bash and zsh.

I don't really get the `alt-p` behavior tbh doesn't really seem that helpful, fish filters on everything you typed. Only way to improve on that is fuzzy filtering or something maybe.
